lib/spack/spack/build_systems/intel.py

```diff
@@ -103,6 +136,742 @@ def _filtered_components(self):
+    # --------------------------------------------------------------------
+    # For MPI, the layout is slightly different than MKL. The prefix will have
+    # to include the sub-arch, which contains bin/, lib/, ..., all without
+    # further arch splits. I_MPI_ROOT, however, must be the package dir.
+    #
+    # FIXME: For MANPATH, need the parent dir.
+    #
+    #   $ ls -lF /opt/intel/compilers_and_libraries_2018.1.163/linux/mpi/
+    #   bin64 -> intel64/bin/
+    #   etc64 -> intel64/etc/
+    #   include64 -> intel64/include/
+    #   lib64 -> intel64/lib/
+    #
+    #   benchmarks/
+    #   binding/
+    #   intel64/
+    #   man/
+    #   test/
+    #
+    #   MPI-2019 preview; Relnotes contain: "File structure clean-up."
+    #   https://software.intel.com/en-us/articles/restoring-legacy-path-structure-on-intel-mpi-library-2019
+    #
+    #   $ ls -lF /opt/intel/compilers_and_libraries_2018.1.163/linux/mpi_2019/
+    #   binding/
+    #   doc/
+    #   imb/
+    #   intel64/
+    #   man/
+    #   test/
+    #
+    # --------------------------------------------------------------------
+    # Spack-external installation of intel-parallel-studio :
+    # --------------------------------------------------------------------
+    #
+    # This is a product bundle whose install dir mostly holds merely symlinks
+    # to components installed in sibling dirs:i
+    #
+    #   $ ls -lF /opt/intel/parallel_studio_xe_2018.1.038/
+    #   advisor_2018		 -> /opt/intel/advisor_2018/
+    #   clck_2018		 -> /opt/intel/clck/2018.1/
+    #   compilers_and_libraries_2018 -> /opt/intel/comp....aries_2018.1.163/
+    #   documentation_2018	 -> /opt/intel/documentation_2018/
+    #   ide_support_2018	 -> /opt/intel/ide_support_2018/
+    #   inspector_2018		 -> /opt/intel/inspector_2018/
+    #   itac_2018		 -> /opt/intel/itac/2018.1.017/
+    #   man		         -> /opt/intel/man/
+    #   samples_2018		 -> /opt/intel/samples_2018/
+    #   vtune_amplifier_2018	 -> /opt/intel/vtune_amplifier_2018/
+    #
+    #   psxevars.csh		 -> ./bin/psxevars.csh*
+    #   psxevars.sh		 -> ./bin/psxevars.sh*
+    #   bin/            - *vars.*sh scripts + sshconnectivity.exp ONLY
+    #
+    #   licensing/
+    #   uninstall*
+    #
+    # The only relevant regular files are *vars.*sh, but those also just churn
+    # through the subordinate vars files of the components.
+    #
+    # --------------------------------------------------------------------
+    #
+    # Note on macOS support, i.e., sys.platform == 'darwin':
+    #
+    # - On macOS, the Spack methods here only include support to integrate an
+    #   externally installed MKL.
+    #
+    # - URLs in child packages will be Linux-specific; macOS download packages
+    #   are located in differently numbered dirs and are named m_*.dmg.
+    #
+    # --------------------------------------------------------------------
+
+    def product_dir(self, product_dir_name, version_glob='_2???.*.*[0-9]',
```

Comment:
I'm worried about this version glob. I wouldn't put it past Intel to go from 2018 to '19 in their directory structures. Could we get away with any sequence of digits and dots? Also, these functions need descriptions of each argument.

Reply:
Right. Looking at my ``/opt/intel``, the form ``version_glob='_*.*.*'`` should do the trick.

All I need this to do is to pick the actual product-release dir ``foo_year.bar.baz`` and not ``foo_year`` which is a compatibiliy directory, with mostly symlinks into the former.

Intel had been wildly rearranging the names over the years, including at one point ``composerxe`` 🙄, followed by ``composer_xe_2011_sp1``.

Reply:
Just wait until it goes from `*.*.*` to `*.*` and then to `*`. Speaking of which, isn't your regex syntax wrong? It matches `_2` followed by 3 characters, followed by anything, ending in a digit. Need to specify what that * modifies, because . matches any character.

Reply:
> Speaking of which, isn't your regex syntax wrong?

Recall that this is a file glob, not a regex.